src/core/geometry/qgspolyhedralsurface.cpp

```diff
@@ -0,0 +1,997 @@
+  {
+    return false;
+  }
+
+  mWkbType = QgsWkbTypes::addZ( mWkbType );
+
+  for ( QgsPolygon *patch : std::as_const( mPatches ) )
+  {
+    patch->addZValue( zValue );
+  }
+  clearCache();
+  return true;
+}
+
+bool QgsPolyhedralSurface::addMValue( double mValue )
+{
+  if ( QgsWkbTypes::hasM( mWkbType ) )
+  {
+    return false;
+  }
+
+  mWkbType = QgsWkbTypes::addM( mWkbType );
+
+  for ( QgsPolygon *patch : std::as_const( mPatches ) )
+  {
+    patch->addMValue( mValue );
+  }
+  clearCache();
+  return true;
+}
+
+bool QgsPolyhedralSurface::dropZValue()
+{
+  if ( !is3D() )
+  {
+    return false;
+  }
+
+  mWkbType = QgsWkbTypes::dropZ( mWkbType );
+  for ( QgsPolygon *patch : std::as_const( mPatches ) )
+  {
+    patch->dropZValue();
+  }
+  clearCache();
+  return true;
+}
+
+bool QgsPolyhedralSurface::dropMValue()
+{
+  if ( !isMeasure() )
+  {
+    return false;
+  }
+
+  mWkbType = QgsWkbTypes::dropM( mWkbType );
+  for ( QgsPolygon *patch : std::as_const( mPatches ) )
+  {
+    patch->dropMValue();
+  }
+  clearCache();
+  return true;
+}
+
+void QgsPolyhedralSurface::swapXy()
+{
+  for ( QgsPolygon *patch : std::as_const( mPatches ) )
+  {
+    patch->swapXy();
+  }
+  clearCache();
+}
+
+QgsMultiSurface *QgsPolyhedralSurface::toCurveType() const
+{
+  QgsMultiSurface *multiSurface = new QgsMultiSurface();
+  multiSurface->reserve( mPatches.size() );
+  for ( const QgsPolygon *polygon : std::as_const( mPatches ) )
+  {
+    multiSurface->addGeometry( polygon->clone() );
+  }
+  return multiSurface;
+}
+
+bool QgsPolyhedralSurface::transform( QgsAbstractGeometryTransformer *transformer, QgsFeedback *feedback )
+{
+  if ( !transformer )
+    return false;
+
+  bool res = true;
+
+  for ( QgsPolygon *patch : std::as_const( mPatches ) )
+  {
+    res = patch->transform( transformer );
+    if ( feedback && feedback->isCanceled() )
+      res = false;
+
+    if ( !res )
+      break;
+  }
+
+  clearCache();
+  return res;
+}
+
+QgsMultiPolygon *QgsPolyhedralSurface::toMultiPolygon() const
+{
+  QgsMultiPolygon *multiPolygon = new QgsMultiPolygon();
+  multiPolygon->reserve( mPatches.size() );
+  for ( const QgsPolygon *polygon : std::as_const( mPatches ) )
+  {
+    multiPolygon->addGeometry( polygon->clone() );
+  }
+  return multiPolygon;
+}
+
+void QgsPolyhedralSurface::filterVertices( const std::function<bool ( const QgsPoint & )> &filter )
+{
+  for ( QgsPolygon *patch : std::as_const( mPatches ) )
+  {
+    patch->filterVertices( filter );
+  }
+
+  clearCache();
+}
+
+void QgsPolyhedralSurface::transformVertices( const std::function<QgsPoint( const QgsPoint & )> &transform )
+{
+  for ( QgsPolygon *patch : std::as_const( mPatches ) )
+  {
+    patch->transformVertices( transform );
+  }
+
+  clearCache();
+}
+
+int QgsPolyhedralSurface::childCount() const
+{
+  return mPatches.count();
+}
+
+QgsAbstractGeometry *QgsPolyhedralSurface::childGeometry( int index ) const
+{
+  return mPatches.at( index - 1 );
```

Comment:
Likely a copy&paste left-over from QgsCurvePolygon::childGeometry()  (shows also a test might be needed for that)
```suggestion
  return mPatches.at( index );
```

Reply:
This probably comes from a copy-paste! I have also added a unit test.